File: moonpoison/heap_priority_queue/875.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) 
    {
        int n=piles.size();
        long int left=1, right=INT_MAX;
        while(left<right)
        {
            long int mid=(left+right)/2;
            int k=0;
            for(auto i:piles)
            {
                if(i%mid!=0) k++;
                k+=i/mid;
            }
            if(k>h) left=mid+1;
            else right=mid;
        }
        return left;
    }
};

File: ice0208/binary-search/875.py
import math


class Solution:
    def getNeedHour(self, piles: list[int], k: int):
        answer = 0
        for pile in piles:
            answer += math.ceil(pile / k)
        return answer

    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        sum_of_piles = sum(piles)
        len_of_piles = len(piles)

        left = math.ceil(sum_of_piles / h)
        right = math.ceil((sum_of_piles - len_of_piles + 1) / (h - len_of_piles + 1))

        while left < right:
            mid = left + (right - left) // 2

            need_hour = self.getNeedHour(piles, mid)
            if need_hour > h:
                left = mid + 1
            else:
                right = mid

        return left


if __name__ == "__main__":
    print(Solution().minEatingSpeed(piles=[3, 6, 7, 11], h=8))
    print(Solution().minEatingSpeed(piles=[30, 11, 23, 4, 20], h=5))
    print(Solution().minEatingSpeed(piles=[30, 11, 23, 4, 20], h=6))
    print(Solution().minEatingSpeed(piles=[2, 2], h=2))
